### PR TITLE
Make qt support fontconfig on Linux

### PR DESCRIPTION
**Pull request by abcfy2**
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/linuxbrew-core/blob/master/CONTRIBUTING.md)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/linuxbrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `brew install <formula>`)?
- [x] Have you included the output of `brew gist-logs <formula>` of the build failure if your PR fixes a build failure. Please quote the exact error message.

-----

### CODE REVIEW

**Comment by abcfy2** [comment]
Make qt could support fontconfig, otherwise will raise this error:

```bash
QFontDatabase: Cannot find font directory /home/linuxbrew/.linuxbrew/lib/fonts.
```

File: Formula/qt.rb [hunk 1]
```diff
@@ -68,6 +69,9 @@ def install
     elsif OS.linux?
       args << "-system-xcb"
       args << "-R#{lib}"
+      args << "-fontconfig"
+      args << "-feature-freetype"
+      args << "-system-freetype
```
Comment (dawidd6) [inline Formula/qt.rb]: You lost a quote at the end.

Comment (abcfy2) [reply Formula/qt.rb]: Oh, thanks. I've force pushed a new commit.

**Comment by iMichka** [comment]
We should rebuild the bottle. Please change `revision 1 unless OS.mac?` to `revision 2 unless OS.mac?`. And remove the sha256 line corresponding to the linux bottle in the bottle block.

**Comment by abcfy2** [comment]
OK. And I find the Ubuntu and Debian configure options, and I will try to use these options and test again. https://buildd.debian.org/status/fetch.php?pkg=qtbase-opensource-src&arch=amd64&ver=5.11.3%2Bdfsg1-4&stamp=1565883980

**Comment by abcfy2** [comment]
I find strange options `-no-avs2` and `-no-avx512`:

https://github.com/Homebrew/linuxbrew-core/blob/master/Formula/qt.rb#L72-L73

I try to remove these options and use `brew install --build-from-source qt`, build failed.

But I just copy all configure options and make manully on my matchine, it works. I don't know why.

And I also look at the Ubuntu build log and configure options: https://launchpadlibrarian.net/445165557/buildlog_ubuntu-eoan-amd64.qtbase-opensource-src_5.12.4+dfsg-4ubuntu1_BUILDING.txt.gz

They did not add `-no-avs2` and `-no-avx512` but can compile very well.

Could anyone help?

**Comment by abcfy2** [comment]
Done. But QtWebEngine build failed, waiting next qt release could support pulseaudio 13.0

**Comment by dawidd6** [comment]
`qt` depends on `gtk`?

**Comment by abcfy2** [comment]
Yes, here is the gtk+ support.

```
Build type: linux-g++ (x86_64, CPU features: aes avx cx16 f16c fsgsbase mmx pclmul popcnt rdrnd sse sse2 sse3 ssse3 sse4.1 sse4.2 sse4)
Compiler: gcc 5.5.0
Configuration: sse2 aesni sse3 ssse3 sse4_1 sse4_2 avx compile_examples enable_new_dtags f16c largefile precompile_header rdrnd shani x86SimdAlways shared rpath release c++11 c++14 concurrent dbus reduce_exports reduce_relocations stl
Build options:
  Mode ................................... release
  Optimize release build for size ........ no
  Building shared libraries .............. yes
  Using C standard ....................... C11
  Using C++ standard ..................... C++14
  Using ccache ........................... no
  Using new DTAGS ........................ yes
  Using precompiled headers .............. yes
  Using LTCG ............................. no
  Target compiler supports:
    SSE .................................. SSE2 SSE3 SSSE3 SSE4.1 SSE4.2
    AVX .................................. AVX
    AVX512 ............................... <none>
    Other x86 ............................ AES F16C RDRAND SHA
    Intrinsics without -mXXX option ...... yes
  Build parts ............................ libs tools
Qt modules and options:
  Qt Concurrent .......................... yes
  Qt D-Bus ............................... yes
  Qt D-Bus directly linked to libdbus .... yes
  Qt Gui ................................. yes
  Qt Network ............................. yes
  Qt Sql ................................. yes
  Qt Testlib ............................. yes
  Qt Widgets ............................. yes
  Qt Xml ................................. yes
Support enabled for:
  Using pkg-config ....................... yes
  udev ................................... yes
  Using system zlib ...................... no
  Zstandard support ...................... no
Qt Core:
  DoubleConversion ....................... yes
    Using system DoubleConversion ........ no
  GLib ................................... yes
  iconv .................................. yes
  ICU .................................... no
  Built-in copy of the MIME database ..... yes
  Tracing backend ........................ <none>
  Logging backends:
    journald ............................. no
    syslog ............................... no
    slog2 ................................ no
  Using system PCRE2 ..................... yes
Qt Network:
  getifaddrs() ........................... yes
  IPv6 ifname ............................ yes
  libproxy ............................... no
  Linux AF_NETLINK ....................... yes
  OpenSSL ................................ yes
    Qt directly linked to OpenSSL ........ no
  OpenSSL 1.1 ............................ yes
  DTLS ................................... yes
  OCSP-stapling .......................... yes
  SCTP ................................... no
  Use system proxies ..................... yes
Qt Gui:
  Accessibility .......................... yes
  FreeType ............................... yes
    Using system FreeType ................ yes
  HarfBuzz ............................... yes
    Using system HarfBuzz ................ no
  Fontconfig ............................. yes
  Image formats:
    GIF .................................. yes
    ICO .................................. yes
    JPEG ................................. yes
      Using system libjpeg ............... yes
    PNG .................................. yes
      Using system libpng ................ yes
  EGL .................................... yes
  OpenVG ................................. no
  OpenGL:
    Desktop OpenGL ....................... yes
    OpenGL ES 2.0 ........................ no
    OpenGL ES 3.0 ........................ no
    OpenGL ES 3.1 ........................ no
    OpenGL ES 3.2 ........................ no
  Vulkan ................................. no
  Session Management ..................... yes
Features used by QPA backends:
  evdev .................................. yes
  libinput ............................... no
  INTEGRITY HID .......................... no
  mtdev .................................. no
  tslib .................................. no
  xkbcommon .............................. yes
  X11 specific:
    XLib ................................. yes
    XCB Xlib ............................. yes
    EGL on X11 ........................... yes
QPA backends:
  DirectFB ............................... no
  EGLFS .................................. yes
  EGLFS details:
    EGLFS OpenWFD ........................ no
    EGLFS i.Mx6 .......................... no
    EGLFS i.Mx6 Wayland .................. no
    EGLFS RCAR ........................... no
    EGLFS EGLDevice ...................... yes
    EGLFS GBM ............................ yes
    EGLFS VSP2 ........................... no
    EGLFS Mali ........................... no
    EGLFS Raspberry Pi ................... no
    EGLFS X11 ............................ yes
  LinuxFB ................................ yes
  VNC .................................... yes
  Mir client ............................. no
  XCB:
    Using system-provided XCB libraries .. yes
    XCB XKB .............................. yes
    XCB XInput ........................... yes
    Native painting (experimental) ....... no
    GL integrations:
      GLX Plugin ......................... yes
        XCB GLX .......................... yes
      EGL-X11 Plugin ..................... yes
Qt Sql:
  SQL item models ........................ yes
Qt Widgets:
  GTK+ ................................... yes
  Styles ................................. Fusion Windows
Qt PrintSupport:
  CUPS ................................... no
Qt Sql Drivers:
  DB2 (IBM) .............................. no
  InterBase .............................. no
  MySql .................................. no
  OCI (Oracle) ........................... no
  ODBC ................................... no
  PostgreSQL ............................. no
  SQLite2 ................................ no
  SQLite ................................. yes
    Using system provided SQLite ......... no
  TDS (Sybase) ........................... no
Qt Testlib:
  Tester for item models ................. yes
Qt SerialBus:
  Socket CAN ............................. yes
  Socket CAN FD .......................... yes
  SerialPort Support ..................... yes
Further Image Formats:
  JasPer ................................. no
  MNG .................................... no
  TIFF ................................... yes
    Using system libtiff ................. yes
  WEBP ................................... yes
    Using system libwebp ................. yes
Qt QML:
  QML network support .................... yes
  QML debugging and profiling support .... yes
  QML tracing JIT support ................ no
  QML sequence object .................... yes
  QML list model ......................... yes
  QML XML http request ................... yes
  QML Locale ............................. yes
  QML delegate model ..................... yes
Qt Quick:
  Direct3D 12 ............................ no
  AnimatedImage item ..................... yes
  Canvas item ............................ yes
  Support for Qt Quick Designer .......... yes
  Flipable item .......................... yes
  GridView item .......................... yes
  ListView item .......................... yes
  TableView item ......................... yes
  Path support ........................... yes
  PathView item .......................... yes
  Positioner items ....................... yes
  Repeater item .......................... yes
  ShaderEffect item ...................... yes
  Sprite item ............................ yes
Qt Scxml:
  ECMAScript data model for QtScxml ...... yes
Qt Gamepad:
  SDL2 ................................... no
Qt 3D:
  Assimp ................................. yes
  System Assimp .......................... no
  Output Qt3D Job traces ................. no
  Output Qt3D GL traces .................. no
  Use SSE2 instructions .................. yes
  Use AVX2 instructions .................. no
  Aspects:
    Render aspect ........................ yes
    Input aspect ......................... yes
    Logic aspect ......................... yes
    Animation aspect ..................... yes
    Extras aspect ........................ yes
Qt 3D Renderers:
  OpenGL Renderer ........................ yes
Qt 3D GeometryLoaders:
  Autodesk FBX ........................... no
Qt Wayland Drivers:
  EGL .................................... yes
  Raspberry Pi ........................... no
  XComposite EGL ......................... yes
  XComposite GLX ......................... yes
  DRM EGL ................................ yes
  libhybris EGL .......................... no
  Shm emulation server buffer integration . yes
Qt Wayland Client Shell Integrations:
  xdg-shell .............................. yes
  xdg-shell unstable v5 (deprecated) ..... yes
  xdg-shell unstable v6 .................. yes
  ivi-shell .............................. yes
  wl-shell (deprecated) .................. yes
Qt Wayland Client ........................ yes
Qt Wayland Compositor .................... yes
Qt Wayland Compositor Layer Plugins:
  VSP2 hardware layer integration ........ no
Qt Bluetooth:
  BlueZ .................................. no
  BlueZ Low Energy ....................... no
  Linux Crypto API ....................... no
  WinRT Bluetooth API (desktop & UWP) .... no
  WinRT advanced bluetooth low energy API (desktop & UWP) . no
Qt Sensors:
  sensorfw ............................... no
Qt Quick Controls 2:
  Styles ................................. Default Fusion Imagine Material Universal
Qt Quick Templates 2:
  Hover support .......................... yes
  Multi-touch support .................... yes
Qt Positioning:
  Gypsy GPS Daemon ....................... no
  WinRT Geolocation API .................. no
Qt Location:
  Qt.labs.location experimental QML plugin . yes
  Geoservice plugins:
    OpenStreetMap ........................ yes
    HERE ................................. yes
    Esri ................................. yes
    Mapbox ............................... yes
    MapboxGL ............................. yes
    Itemsoverlay ......................... yes
QtXmlPatterns:
  XML schema support ..................... yes
Qt Multimedia:
  ALSA ................................... yes
  GStreamer 1.0 .......................... no
  GStreamer 0.10 ......................... no
  Video for Linux ........................ yes
  OpenAL ................................. no
  PulseAudio ............................. yes
  Resource Policy (libresourceqt5) ....... no
  Windows Audio Services ................. no
  DirectShow ............................. no
  Windows Media Foundation ............... no
Qt Tools:
  QDoc ................................... no
Qt WebEngineCore:
  Qt WebEngine Widgets ................... yes
  Qt WebEngine Qml ....................... yes
  Embedded build ......................... no
  Full debug information ................. no
  Pepper Plugins ......................... yes
  Printing and PDF ....................... yes
  Proprietary Codecs ..................... yes
  Spellchecker ........................... yes
  Native Spellchecker .................... no
  WebRTC ................................. yes
  Use System Ninja ....................... no
  Geolocation ............................ yes
  WebChannel support ..................... yes
  Use v8 snapshot ........................ yes
  Kerberos Authentication ................ no
  Extensions ............................. no
  Support qpa-xcb ........................ yes
  Use ALSA ............................... yes
  Use PulseAudio ......................... yes
  Optional system libraries used:
    re2 .................................. no
    icu .................................. no
    libwebp, libwebpmux and libwebpdemux . no
    opus ................................. no
    ffmpeg ............................... no
    libvpx ............................... yes
    snappy ............................... yes
    glib ................................. yes
    zlib ................................. no
    minizip .............................. no
    libevent ............................. no
    jsoncpp .............................. no
    protobuf ............................. no
    libxml2 and libxslt .................. no
    lcms2 ................................ no
    png .................................. yes
    JPEG ................................. no
    harfbuzz ............................. no
    freetype ............................. yes
  Required system libraries:
    fontconfig ........................... yes
    dbus ................................. yes
    nss .................................. yes
    khr .................................. yes
    glibc ................................ yes
  Required system libraries for qpa-xcb:
    x11 .................................. yes
    libdrm ............................... yes
    xcomposite ........................... yes
    xcursor .............................. yes
    xi ................................... yes
    xtst ................................. yes
Qt WebEngineQml:
  UI Delegates ........................... yes
  Test Support ........................... no
```

And also, I find the ubuntu build support gtk+: https://launchpadlibrarian.net/445165557/buildlog_ubuntu-eoan-amd64.qtbase-opensource-src_5.12.4+dfsg-4ubuntu1_BUILDING.txt.gz

**Comment by dawidd6** [comment]
```
Error: The `brew link` step did not complete successfully
```

Something wrong with `pulseaudio`

**Comment by abcfy2** [comment]
It seems that an issue about `pulseaudio`.

**Comment by dawidd6** [comment]
Standalone install of `pulseaudio` works just right. Maybe some files are conflicting with `qt` or anything else.

**Comment by abcfy2** [comment]
```
The formula built, but is not symlinked into /home/linuxbrew/.linuxbrew
Could not symlink share/glib-2.0/schemas/gschemas.compiled
Target /home/linuxbrew/.linuxbrew/share/glib-2.0/schemas/gschemas.compiled
already exists. You may want to remove it:
  rm '/home/linuxbrew/.linuxbrew/share/glib-2.0/schemas/gschemas.compiled'

To force the link and overwrite all conflicting files:
  brew link --overwrite pulseaudio

To list all files that would be deleted:
  brew link --overwrite --dry-run pulseaudio

Possible conflicting files are:
mkdir -p /home/linuxbrew/.linuxbrew/etc/pulse
/home/linuxbrew/.linuxbrew/share/glib-2.0/schemas/gschemas.compiled
mkdir -p /home/linuxbrew/.linuxbrew/share/zsh
mkdir -p /home/linuxbrew/.linuxbrew/share/zsh/site-functions
==> Caveats
Warning: pulseaudio provides a launchd plist which can only be used on macOS!
You can manually execute the service instead with:
  pulseaudio
==> Summary
/home/linuxbrew/.linuxbrew/Cellar/pulseaudio/13.0: 224 files, 6.2MB
```

**Comment by dawidd6** [comment]
https://stackoverflow.com/a/56317158

**Comment by abcfy2** [comment]
I think we could delete `gschemas.compiled` after `pulseaudio` compiled ?

```bash
$ find `brew --prefix` -name gschemas.compiled
/home/linuxbrew/.linuxbrew/share/glib-2.0/schemas/gschemas.compiled
/home/linuxbrew/.linuxbrew/Cellar/pulseaudio/13.0/share/glib-2.0/schemas/gschemas.compiled
```

You can see after `glib` installed, `/home/linuxbrew/.linuxbrew/share/glib-2.0/schemas/gschemas.compiled` will be generated. But `pulseaudio` generated another `gschemas.compiled`

Maybe we can do like this:

```diff
diff --git a/Formula/pulseaudio.rb b/Formula/pulseaudio.rb
index 2ebbe57..23e99e5 100644
--- a/Formula/pulseaudio.rb
+++ b/Formula/pulseaudio.rb
@@ -34,6 +34,7 @@ class Pulseaudio < Formula
     depends_on "expat"
     depends_on "glib"
     depends_on "libcap"
+    depends_on "dbus"
 
     # Depends on XML::Parser
     # Using the host's Perl interpreter to install XML::Parser fails when using brew's glibc.
@@ -88,6 +89,7 @@ class Pulseaudio < Formula
       system "./configure", *args
     end
     system "make", "install"
+    system "rm", "-f", "#{prefix}/share/glib-2.0/schemas/gschemas.compiled"
   end
 
   plist_options :manual => "pulseaudio"
```

And it works:

```bash
$ brew install -s pulseaudio 
==> Downloading https://www.freedesktop.org/software/pulseaudio/releases/pulseaudio-13.0.tar.xz
Already downloaded: /home/ygc/.cache/Homebrew/downloads/a9d9de03a3e228cd1ccbca18a37d64a80af4ace1d91a1ef1a62d4d72962c5452--pulseaudio-13.0.tar.xz
==> Downloading https://cpan.metacpan.org/authors/id/T/TO/TODDR/XML-Parser-2.44.tar.gz
Already downloaded: /home/ygc/.cache/Homebrew/downloads/3948f45448ce280a3858f863934ea711a56695d7a55e179df695b809722c0767--XML-Parser-2.44.tar.gz
==> perl Makefile.PL INSTALL_BASE=/home/linuxbrew/.linuxbrew/Cellar/pulseaudio/13.0/libexec
==> make PERL5LIB=/home/linuxbrew/.linuxbrew/Cellar/pulseaudio/13.0/libexec/lib/perl5 CC=gcc-5
==> make install
==> ./configure --disable-silent-rules --prefix=/home/linuxbrew/.linuxbrew/Cellar/pulseaudio/13.0 --disable-neon-opt --disable-nls --disable-x11 --with-database=simple
==> make install
==> rm -f -r /home/linuxbrew/.linuxbrew/Cellar/pulseaudio/13.0/share/glib-2.0/schemas/gschemas.compiled
==> Caveats
Warning: pulseaudio provides a launchd plist which can only be used on macOS!
You can manually execute the service instead with:
  pulseaudio
==> Summary
🍺  /home/linuxbrew/.linuxbrew/Cellar/pulseaudio/13.0: 228 files, 5.9MB, built in 1 minute 28 seconds
```

No errors.

**Comment by abcfy2** [comment]
Hi @iMichka . Please review this PR, and let me know if there are any problems. Thanks.

File: Formula/qt.rb [hunk 2]
```diff
@@ -65,13 +68,30 @@ def install
...
+      args << "-platform"
+      args << "linux-g++"
       args << "-system-xcb"
-      args << "-R#{lib}"
-      # https://bugreports.qt.io/projects/QTBUG/issues/QTBUG-71564
-      args << "-no-avx2"
-      args << "-no-avx512"
+      args << "-openssl"
+      args << "-qpa"
+      args << "xcb"
```
Comment (iMichka) [inline Formula/qt.rb]: There is a system-xcb flag just above. Now we would have 3 xcb flags. Why?

File: Formula/qt.rb [hunk 3]
```diff
@@ -65,13 +68,30 @@ def install
...
+      args << "-xcb"
+      args << "-glib"
+      args << "-dbus-linked"
+      args << "-no-directfb"
+      args << "-no-use-gold-linker"
+      args << "-accessibility"
```
Comment (iMichka) [inline Formula/qt.rb]: Regarding these 3 flags, can they be added to the mac repo first?

File: Formula/qt.rb (hunk 2 above)
Comment (abcfy2) [reply Formula/qt.rb]: Just from Ubuntu/Debian: https://launchpadlibrarian.net/445165557/buildlog_ubuntu-eoan-amd64.qtbase-opensource-src_5.12.4+dfsg-4ubuntu1_BUILDING.txt.gz

File: Formula/qt.rb (hunk 3 above)
Comment (abcfy2) [reply Formula/qt.rb]: I have no mac, I don't know if these flags could work on mac.

Comment (iMichka) [reply Formula/qt.rb]: Then I need to: test these on mac / find the documentation about these.
We want to keep the mac and linux formulae as much as possible in sync.

File: Formula/qt.rb [hunk 4]
```diff
@@ -28,6 +27,11 @@ class Qt < Formula
 
   unless OS.mac?
     depends_on "fontconfig"
+    depends_on "nss"
+    depends_on "openssl@1.1"
```
Comment (iMichka) [inline Formula/qt.rb]: `brew deps qt` tells me that `qt` already depends on `openssl@1.1`. This is probably not needed.

File: Formula/qt.rb (hunk 2 above)
Comment (iMichka) [reply Formula/qt.rb]: Weird. I already had a hard time with these flags. Does adding these flags fix a bug for you/enable more features? If not, I do not see why we would add them. I would rather prefer we do not "blindly" follow all the flags added by ubuntu/debian.

**Comment by iMichka** [comment]
Regarding this PR: it started with a fix for fontconfig, and it now diverged to: lets try to enable as many flags as possible, or at least try to follow what ubuntu/debian did.

Please note that ubuntu/debian have way more developers to support all these configurations ... we do not have that manpower. So we would rather have less flags. And we try to follow as much as possible https://github.com/Homebrew/homebrew-core. Also, I would like to not increase the dependency list too much.

Could we come back to a smaller PR with less things?

If one specific flag fixes a bug for you, I would be willing to review it. But let's try to focus on fontconfig first. I need smaller changes to be able to review this in a timely manner.

**Comment by iMichka** [comment]
> If one specific flag fixes a bug for you, I would be willing to review it. But let's try to focus on fontconfig first. I need smaller changes to be able to review this in a timely manner.

As this needs to be rebased anyway, I'll close this. Please open a new pull request with one change at at time, so we can review it one by one and focus on each item.

If the build fails, we can try to help you, but we have limited manpower, so we may not be able to enable everything at once.

**Pull request closed without merging**